### ISSUE TITLE
Semantic segmentation

### ISSUE THREAD

**Issue by haiton266** [phiyodr/dacl10k-toolkit#5]
I have issue when convert this annotation to semantic segmentation because I don't know what which class is above which class ... I test so I see current code only convert to 19 channels but I only want 1 channel. Can you suggest more for me, please? Thank you!

**Comment by phiyodr** [comment]
Hi!
For a quick-fix just return the one channel from `mask` you are interested in. 
Apply slicing here: https://github.com/phiyodr/dacl10k-toolkit/blob/9d4f9e8d2f07fbb3a520d72f80308ac80244e8f4/dacl10k/dacl10kdataset.py#L164, e. g. using `0` for "Crack" (https://github.com/phiyodr/dacl10k-toolkit/blob/9d4f9e8d2f07fbb3a520d72f80308ac80244e8f4/dacl10k/dacl10kdataset.py#L44).

**Comment by phiyodr** [comment]
We use _Multi-label Semantic Segmentation_ since each pixel can be associated with multiple labels. Therefore, we use a total of 19 binary masks—one mask for each label.

If you want to use only one mask ("Multi-class Semantic Segmentation"), you would need to map each possible pixel combination (e.g., [1, 0, ..., 0] or [1, 1, 0, ..., 0]) to a unique identifier. This is not implemented, and I would not recommend it, since it makes it harder for the model to learn the mapping.

If you are only interested in one label or in labels which never overlap, of course you can use that. Therefore you need to apply slicing—as mentioned above—and index mapping.  

For  _Multi-label Semantic Segmentation_  you can use tools like https://github.com/qubvel-org/segmentation_models.pytorch.

**Comment by phiyodr** [comment]
As far as I know they have no demo for the multilabel case. The dacl10k target mask has the shape `(19, 512, 512)` with 19 multilabel classes. You need 
1. ... to adjust the model to support 19 classes, e.g. adjusting `classes` in https://smp.readthedocs.io/en/latest/models.html#deeplabv3plus
2. ... a loss that supports multilabel mode, e.g. https://smp.readthedocs.io/en/latest/losses.html#diceloss (`mode (str) – Loss mode ‘binary’, ‘multiclass’ or ‘multilabel’`)

**Comment by wangxin1999** [comment]
> Thank you! It worked

hi, could you share the code?

**Comment by haiton266** [comment]
I refer from example https://github.com/qubvel-org/segmentation_models.pytorch/blob/main/examples/camvid_segmentation_multiclass.ipynb and modify it. You can refer this code:
```python
import pytorch_lightning as pl
import segmentation_models_pytorch as smp
import torch
from torch.optim import lr_scheduler
import wandb

class CamVidModel(pl.LightningModule):
    def __init__(self, arch, encoder_name, in_channels, out_classes, **kwargs):
        super().__init__()
        self.model = smp.create_model(
            arch,
            encoder_name=encoder_name,
            in_channels=in_channels,
            classes=out_classes,
            **kwargs,
        )

        # Preprocessing parameters for image normalization
        params = smp.encoders.get_preprocessing_params(encoder_name)
        self.number_of_classes = out_classes
        self.register_buffer("std", torch.tensor(params["std"]).view(1, 3, 1, 1))
        self.register_buffer("mean", torch.tensor(params["mean"]).view(1, 3, 1, 1))

        # Loss function for multi-label segmentation
        self.loss_fn = smp.losses.DiceLoss(mode="multilabel", from_logits=True)

        # Step metrics tracking
        self.training_step_outputs = []
        self.validation_step_outputs = []
        self.test_step_outputs = []

    def forward(self, image):
        # Normalize image
        image = (image - self.mean) / self.std
        mask = self.model(image)
        return mask

    def shared_step(self, batch, stage):
        image, mask = batch

        # Ensure that image dimensions are correct
        assert image.ndim == 4  # [batch_size, channels, H, W]

        # Ensure mask dimensions for multilabel segmentation
        assert mask.ndim == 4  # [batch_size, num_classes, H, W]

        # Predict mask logits
        logits_mask = self.forward(image)

        assert (
            logits_mask.shape[1] == self.number_of_classes
        )  # [batch_size, num_classes, H, W]

        # Compute loss
        loss = self.loss_fn(logits_mask, mask)

        # Apply sigmoid activation for multilabel probabilities
        prob_mask = torch.sigmoid(logits_mask)

        # Binarize predictions using a threshold
        pred_mask = (prob_mask > 0.5).int()

        # Log loss to wandb
        self.log(f"{stage}/{stage}_loss", loss, on_step=True, on_epoch=True, prog_bar=True)

        # Compute true positives, false positives, false negatives, and true negatives
        tp, fp, fn, tn = smp.metrics.get_stats(
            pred_mask, mask.int(), mode="multilabel", num_classes=self.number_of_classes
        )

        return {
            "loss": loss,
            "tp": tp,
            "fp": fp,
            "fn": fn,
            "tn": tn,
        }
    # This code isssue, need to check it
    def shared_epoch_end(self, outputs, stage):
        # Aggregate step metrics across all batches
        tp = torch.cat([x["tp"] for x in outputs], dim=0)  # [num_samples, num_classes]
        fp = torch.cat([x["fp"] for x in outputs], dim=0)
        fn = torch.cat([x["fn"] for x in outputs], dim=0)
        tn = torch.cat([x["tn"] for x in outputs], dim=0)

        # Refer: https://smp.readthedocs.io/en/latest/metrics.html#segmentation_models_pytorch.metrics.functional.iou_score
        # Metric for standard: https://openaccess.thecvf.com/content/WACV2024W/DACL/papers/Flotzinger_Dacl-Challenge_Semantic_Segmentation_During_Visual_Bridge_Inspections_WACVW_2024_paper.pdf (3.2)

        # Calculate overall metrics
        m_iou = smp.metrics.iou_score(tp, fp, fn, tn, reduction="macro")

        # Log overall metrics
        metrics = {
            f"{stage}/{stage}_mean_iou": m_iou.item(),
        }
        self.log_dict(metrics, prog_bar=True)

    def training_step(self, batch, batch_idx):
        train_loss_info = self.shared_step(batch, "train")

        # Log learning rate at this step
        lr = self.trainer.optimizers[0].param_groups[0]['lr']  # Get learning rate from the optimizer
        self.log("train/lr", lr, on_step=True, on_epoch=False, prog_bar=True)

        self.training_step_outputs.append(train_loss_info)
        return train_loss_info

    def on_train_epoch_end(self):
        self.shared_epoch_end(self.training_step_outputs, "train")
        self.training_step_outputs.clear()

    def validation_step(self, batch, batch_idx):
        valid_loss_info = self.shared_step(batch, "valid")
        self.validation_step_outputs.append(valid_loss_info)
        return valid_loss_info

    def on_validation_epoch_end(self):
        self.shared_epoch_end(self.validation_step_outputs, "valid")
        self.validation_step_outputs.clear()

    def test_step(self, batch, batch_idx):
        test_loss_info = self.shared_step(batch, "test")
        self.test_step_outputs.append(test_loss_info)
        return test_loss_info

    def on_test_epoch_end(self):
        self.shared_epoch_end(self.test_step_outputs, "test")
        self.test_step_outputs.clear()

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-4)
        scheduler = lr_scheduler.CosineAnnealingLR(optimizer, T_max=50, eta_min=1e-5)
        return {
            "optimizer": optimizer,
            "lr_scheduler": {
                "scheduler": scheduler,
                "interval": "step",
                "frequency": 1,
            },
        }
```
